### PR TITLE
can't save page with listing block and standard caching rules (due to etag not showing lock)

### PR DESCRIPTION
**Pull request by JeffersonBledsoe**
# Problem

if you use volto and a listing block with current folder contents and caching turned on then then when you try to save it will tell you the page is locked with a 403 and you can't save.
This is when you have applied the standard caching rules to restapi, in particular moderate (if not modified) caching to content api

# Steps to reproduce

1. turn on caching and ensure content has moderate caching (304 not modified)
2. create a page with listing block showing page contents
3. save
4. reedit
5. save again

# Expected behaviour

You can save it and the content will be unlocked

# Actual behaviour

You get an error saying the content is already locked.

# Cause
Prior to this PR, the `object locked` status code within the etag header was a `1` if the object was locked by another user, `0` if the object was locked by the current user or it was unlocked. 
This didn't seem to cause any issues with plone classic.

However in volto any ploneapi request for content will return the lock token in the info in the response and if none is there then it assumes the content is unlocked.

In this particular case 
1. the first request for content might cause the api to be cached with an etag with 0 for unlocked.
2. then an api request is made to lock and the lock token stored in volto state
3. then the block to list the folder contents causes another content api call (there is no seperate api for this). Since the user is the one that locked the content the etag is the same so the local cached api response is used which indicates the content is unlocked. Volto decides to clear the lock token and believes the content is no unlocked.
4. volto doesn't send the lock token with the save and so the save fails

Really the content should not be cached if the user has logged in and the lock token has changed since this changes the response you get.

# Proposed solution

This PR changes the `ObjectLocked` value in etag to instead return the current lock token if the object is locked. This fixes the false-positive cache hit between locking and unlocking. The lock key is returned to avoid any edge-case false-positive hits between users locking the content and returning to that locked content. The lock key does not contain any sensitive information

# Other solutions considered
- could perhaps not show the lock in the etag when the user is not logged in or doesn't have permission to edit. This would result in more cache hits. 
- you could just ensure that '1' is in the etag even if you locked it. However there is an edge case where you edit the content. browse away. someone unlocks and then locks it again. Then you come back and it will think you locked it and not the other person leading to a similar bug.
- maybe could make volto smarter to not change the lock token on subsequent content calls? really it needs a better story about what happens if someone changes the lock halfway through an edit.

Tests failing due to needing to add the locking behaviour to the mocked content object. I'll get this resolved

### CODE REVIEW

**Comment by jensens** [comment]
pre-commit.ci autofix

**Comment by fredvd** [comment]
@JeffersonBledsoe Thank you for creating this PR. Nasty edge case.  Root cause analysis:  plone.restapi implemented a different locking communication than what is/was there already? 

What I'm missing in your analysis:  what (if?) are the implications for Classic-UI Plone with this change? I'm only reading disucssion around what is going wrong with Volto because it decided to leave out the Locking status in the Etag.

**Comment by fredvd** [comment]
@mauritsvanrees Can we discuss this later today/tomorrow?

**Comment by fredvd** [comment]
@jensens What is your opinion? You're adding me as a reviewer, but IMHO you've worked more on caching support than me the last 12-18 months.  

Also cc:eing @ericof here.

**Comment by jensens** [comment]
> @jensens What is your opinion?

Well, I am not that deep into details of volto+caching with logged in users. AFAIK @ericof worked on this topic in the last months. I would have to dig deeper into it, lacking time currently to do so.

**Comment by fredvd** [comment]
Something else. I'm almost certain that stealing locks was limited to Managers or Site Administrators with the correct permissions only assigned for them in the past.  Since when did it change to all Editors?  Or am I mistaken with something else?  Maurits didn't see any permission checks in the code when he scanned it this morning.

**Comment by JeffersonBledsoe** [comment]
>If the listing block is resetting the Locking info, that's a bug that should be fixed in the listing block first.

The reason the listing block is mentioned is because there's an optimisation in Volto where a listing with an empty query will use the "items" key from the `get` request of the current content (generated from a query of objects in parents with a depth of 1) rather than performing another query. As this re-uses the same get request that was done to initially fetch the object for editing, the cached response is returned.

> With most of our caching setups for Classic-UI we don't have this problem as logged in editors have an __ac cookie set and the default varnish configuration doesn't and shouldn't cached any logged in requests for html and ajax requests etc. (we strip in the plone.recipe.varnish recipe vcl __ac from all 'resource' types.

Agreed that this isn't an issue in Plone Classic. As you've identified though, the REST API uses a `Lock-Token` header to verify whether the current user has permission rather than the normal lock permission checking.

https://github.com/plone/plone.restapi/blob/b15ddc03e1add4197458f8ba55551e62a0d58b3d/src/plone/restapi/services/content/update.py#L14-L16

https://github.com/plone/plone.restapi/blob/b15ddc03e1add4197458f8ba55551e62a0d58b3d/src/plone/restapi/services/locking/__init__.py#L63-L76

**Comment by djay** [comment]
> Agreed that this isn't an issue in Plone Classic. As you've identified though, the REST API uses a Lock-Token header to verify whether the current user has permission rather than the normal lock permission checking.

to be 100% clear is the only implication for classic is that moderate caching might result in more misses in the very rare case a lock is stolen or the lock is just changed because someone else is editing it. I don't see this making any significant difference to caching performance. There is an argument that classic should not have locking in the etag for anon users since it doesn't change the response but this is not the case with the restapi and its easier to keep both the same.

**Comment by jensens** [comment]
Something I found:
Are you aware there are caching rules for endpoints in place?
https://github.com/plone/documentation/blob/6-dev/docs/deployment/caching/restapi.md
The document could be outdated, as there were endpoints added after I wrote it.
The caching rules for endpoints are defined in ZCML side by side with their registration like here https://github.com/plone/documentation/blob/6-dev/docs/deployment/caching/restapi.md 
but those are missing here 
https://github.com/plone/plone.restapi/blob/master/src/plone/restapi/services/locking/configure.zcml

**Comment by djay** [comment]
@jensens This uses the caching rules. The issue arises when using caching rules because the content endpoint changes in a way not reflected by the etag so a cached version is used when it shouldn't

**Comment by djay** [comment]
@jensens can we get this merged? This breaks your site so you are not able to save any content if you have applied the default recommended caching rules for restapi.

File: plone/app/caching/operations/etags.py [hunk 1]
```diff
@@ -157,12 +157,16 @@ def __init__(self, published, request):
...
         context = getContext(self.published)
-        context_state = queryMultiAdapter(
-            (context, self.request), name="plone_context_state"
-        )
-        if context_state is None:
-            return
-        return "1" if context_state.is_locked() else "0"
+        lock = queryMultiAdapter((context, self.request), name="plone_lock_info")
+
+        if not lock or not lock.lock_info():
```
Comment (davisagli) [inline plone/app/caching/operations/etags.py:162]:
```suggestion
        if not lock:
```

(Avoid calling lock_info() twice)

**Comment by davisagli** [comment]
@jenkins-plone-org please run jobs

**Comment by davisagli** [comment]
Note: the failing circular dependencies check is unrelated to this repository and PR.

**Review by davisagli — APPROVED** [review]

**Pull request merged**